File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/full/EventFull.java
/*
 * Copyright (c) Kacper Ziubryniewicz 2020-1-6
 */

package pl.szczodrzynski.edziennik.data.db.full;

import pl.szczodrzynski.edziennik.data.db.entity.Event;
import pl.szczodrzynski.edziennik.data.db.entity.Metadata;

public class EventFull extends Event {
    public String typeName = "";
    public int typeColor = -1;

    public String teacherFullName = "";

    public String subjectLongName = "";
    public String subjectShortName = "";

    public String teamName = "";
    public String teamCode = null;

    // metadata
    public boolean seen;
    public boolean notified;
    public long addedDate;

    public EventFull() {}

    public EventFull(Event event) {
        super(
                event.profileId,
                event.id,
                event.eventDate.clone(),
                event.startTime == null ? null : event.startTime.clone(),
                event.topic,
                event.color,
                event.type,
                event.addedManually,
                event.teacherId,
                event.subjectId,
                event.teamId
        );

        this.sharedBy = event.sharedBy;
        this.sharedByName = event.sharedByName;
        this.blacklisted = event.blacklisted;
    }
    public EventFull(EventFull event) {
        super(
                event.profileId,
                event.id,
                event.eventDate.clone(),
                event.startTime == null ? null : event.startTime.clone(),
                event.topic,
                event.color,
                event.type,
                event.addedManually,
                event.teacherId,
                event.subjectId,
                event.teamId
        );

        this.sharedBy = event.sharedBy;
        this.sharedByName = event.sharedByName;
        this.blacklisted = event.blacklisted;
        this.typeName = event.typeName;
        this.typeColor = event.typeColor;
        this.teacherFullName = event.teacherFullName;
        this.subjectLongName = event.subjectLongName;
        this.subjectShortName = event.subjectShortName;
        this.teamName = event.teamName;
        this.teamCode = event.teamCode;
        this.seen = event.seen;
        this.notified = event.notified;
        this.addedDate = event.addedDate;
    }

    public EventFull(Event event, Metadata metadata) {
        this(event);

        this.seen = metadata.seen;
        this.notified = metadata.notified;
        this.addedDate = metadata.addedDate;
    }

    public int getColor() {
        return color == -1 ? typeColor : color;
    }

    @Override
    public String toString() {
        return "EventFull{" +
                "profileId=" + profileId +
                ", id=" + id +
                ", eventDate=" + eventDate +
                ", startTime=" + startTime +
                ", topic='" + topic + '\'' +
                ", color=" + color +
                ", type=" + type +
                ", addedManually=" + addedManually +
                ", sharedBy='" + sharedBy + '\'' +
                ", sharedByName='" + sharedByName + '\'' +
                ", blacklisted=" + blacklisted +
                ", teacherId=" + teacherId +
                ", subjectId=" + subjectId +
                ", teamId=" + teamId +
                ", typeName='" + typeName + '\'' +
                ", teacherFullName='" + teacherFullName + '\'' +
                ", subjectLongName='" + subjectLongName + '\'' +
                ", subjectShortName='" + subjectShortName + '\'' +
                ", teamName='" + teamName + '\'' +
                ", seen=" + seen +
                ", notified=" + notified +
                ", addedDate=" + addedDate +
                '}';
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/full/EventFull.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2020-4-25.
 */
package pl.szczodrzynski.edziennik.data.db.full

import androidx.room.Ignore
import androidx.room.Relation
import pl.szczodrzynski.edziennik.data.db.entity.Event
import pl.szczodrzynski.edziennik.data.db.entity.Metadata
import pl.szczodrzynski.edziennik.data.db.entity.Note
import pl.szczodrzynski.edziennik.data.db.entity.Noteable
import pl.szczodrzynski.edziennik.ui.search.Searchable
import pl.szczodrzynski.edziennik.utils.html.BetterHtml
import pl.szczodrzynski.edziennik.utils.models.Date
import pl.szczodrzynski.edziennik.utils.models.Time

class EventFull(
        profileId: Int, id: Long, date: Date, time: Time?,
        topic: String, color: Int?, type: Long,
        teacherId: Long, subjectId: Long, teamId: Long, addedDate: Long = System.currentTimeMillis()
) : Event(
        profileId, id, date, time,
        topic, color, type,
        teacherId, subjectId, teamId, addedDate
), Searchable<EventFull>, Noteable {
    constructor(event: Event, metadata: Metadata? = null) : this(
            event.profileId, event.id, event.date, event.time,
            event.topic, event.color, event.type,
            event.teacherId, event.subjectId, event.teamId, event.addedDate) {
        event.let {
            addedManually = it.addedManually
            sharedBy = it.sharedBy
            sharedByName = it.sharedByName
            blacklisted = it.blacklisted
            isDownloaded = it.isDownloaded
            homeworkBody = it.homeworkBody
            attachmentIds = it.attachmentIds
            attachmentNames = it.attachmentNames
        }
        metadata?.let {
            seen = it.seen
            notified = it.notified
        }
    }

    var typeName: String? = null
    var typeColor: Int? = null

    var teacherName: String? = null
    var subjectLongName: String? = null
    var subjectShortName: String? = null
    var teamName: String? = null
    var teamCode: String? = null

    @delegate:Ignore
    @delegate:Transient
    val topicHtml by lazy {
        BetterHtml.fromHtml(context = null, topic, nl2br = true)
    }

    @delegate:Ignore
    @delegate:Transient
    val bodyHtml by lazy {
        homeworkBody?.let {
            BetterHtml.fromHtml(context = null, it, nl2br = true)
        }
    }

    @Ignore
    @Transient
    override var searchPriority = 0

    @Ignore
    @Transient
    override var searchHighlightText: String? = null

    @delegate:Ignore
    @delegate:Transient
    override val searchKeywords by lazy {
        listOf(
            listOf(topicHtml.toString(), bodyHtml?.toString()),
            attachmentNames,
            listOf(subjectLongName),
            listOf(teacherName),
            listOf(sharedByName),
        )
    }

    override fun compareTo(other: Searchable<*>): Int {
        if (other !is EventFull)
            return 0
        return when {
            // ascending sorting
            searchPriority > other.searchPriority -> 1
            searchPriority < other.searchPriority -> -1
            // ascending sorting
            date > other.date -> 1
            date < other.date -> -1
            // ascending sorting
            (time?.value ?: 0) > (other.time?.value ?: 0) -> 1
            (time?.value ?: 0) < (other.time?.value ?: 0) -> -1
            // ascending sorting
            addedDate > other.addedDate -> 1
            addedDate < other.addedDate -> -1
            else -> 0
        }
    }

    // metadata
    var seen = false
    var notified = false

    val eventColor
        get() = color ?: typeColor ?: 0xff2196f3.toInt()

    @Relation(parentColumn = "eventId", entityColumn = "noteOwnerId", entity = Note::class)
    override lateinit var notes: MutableList<Note>
    override fun getNoteType() = Note.OwnerType.EVENT
    override fun getNoteOwnerProfileId() = profileId
    override fun getNoteOwnerId() = id
}
